enhancements/console/dynamic-plugins.md

```diff
@@ -0,0 +1,392 @@
+| `flags?: Partial<{ required: string[]; disallowed: string[]; }>` | Reference to the flags used to trigger the enablement of this extension. |
+
+### Lazy Loading
+
+Console uses
+[code splitting](https://webpack.js.org/guides/code-splitting/) to only load
+JavaScript code when needed. Likewise, dynamic plugin code will only load when
+needed. For instance, if the plugin contributes a list page, the plugin code
+for that list page only loads when the user visits the page. This improves
+performance and guards against bugs in plugins the user is not actively using.
+
+Console performs feature detection on initial load. This is used today to
+enable or disable parts of the UI based on features in the cluster and is
+typically driven by the resources found during API discovery. Dynamic plugins
+will also be able to declare console feature flags and will only load if the
+flag is enabled. For instance, the KubeVirt plugin will only load if the
+`VirtualMachine` resource is present.
+
+### Refreshing When a Plugin Updates
+
+Dynamic plugin code won't update in the browser automatically when a plugin is
+added, removed, or updated. The user must refresh the browser.
+
+The console backend will expose an endpoint that lists the available plugins
+and the current plugin version from each manifest. The frontend will poll this
+endpoint periodically to detect changes. When a change is detected, console
+shows a message in a toast notification indicating that there is an update
+available and the user must refresh their browser to see the changes. (We won't
+refresh the page automatically to avoid possibly losing data if the user is
+entering something into a form or the YAML editor.)
+
+### Error Handling
+
+Console will guard against runtime errors in plugins. All plugin components
+will be surrounded by [React error boundaries](https://reactjs.org/docs/error-boundaries.html).
+This prevents an uncaught error from causing the application to white-screen
+and break. If a plugin service is unavailable, console will not load the plugin
+and show a message in the notification drawer to let users know.
+
+Additionally, the console can contain a `?disable-plugins` query parameter with
+a comma separated list of plugin names. When this parameter is present, the
+corresponding dynamic plugins are not loaded. If `all` is passed, all dynamic
+plugins will be disabled.
+
+### Risks and Mitigations
+
+**Risk**: Patternfly updates with breaking API changes can break plugins.
+
+**Mitigation**: Plugins can declare a semver range for compatibility. The core
+console team will coordinate breaking changes with the plugin teams so that a
+new operator version will be available with fixes when the corresponding
+OpenShift version GAs. We will not ship any breaking Patternfly changes in a
+z-stream.
+
+**Risk**: Plugins are not sandboxed and can make any API request on behalf of the user.
+
+**Mitigation**: Plugins are not enabled by default. A cluster admin must opt-in.
```

Comment:
- they cannot be audited
- they can install backdoors (e.g. via service-accounts)
- they can risk the cluster by bad behaviour.

I would expect a discusion about alternatives, why we don't care about security at all.

Reply:
the same can be said for operators, as well as workloads that a cluster admin chooses to grant admin permissions to.

obviously the set of people who can install a plugin on the cluster needs to be limited and the cluster-admin needs to understand these risks when choosing to install/enable a plugin.

The initial intent for this feature is that it is only used by internal redhat components.  It may or may not be extended to third parties in the future, but again this falls into the category of "if the cluster admin wants to hand out access to the cluster by allowing code onto the cluster, that is their decision"

We should of course ensure the implications of this are clearly documented for admins who need to make these decisions.

Reply:
As @bparees mention we are aiming to:
- have a simply mechanism for the cluster admin to enable/disable the ConsolePlugin
- only redhat owned components will have permissions to create ConsolePlugin CRs
- use the ConsolePlugins for porting the static plugins(knative, helm, insights, ...) that we currently have in the console repo into dynamic plugins. The point is to decouple them from the core console, so they can have their own release cycle.
- Also as part of the work we with @vojtechszocs [demo DynamicPlugin](https://github.com/openshift/console/pull/7471/files), which besides Dockerfile contains manifests for creating namespace, deployment and service. Service annotation contains `service.alpha.openshift.io/serving-cert-secret-name: console-serving-cert` that should inject `console-serving-cert` secret into the ConsolePlugin pod and used for serving plugins assets.

We are not aiming to have this production ready for 4.7, we just want to give teams that own console's static plugins a way how to decouple from console. We do care about the security, but since with this feature we are not aiming for production yet, it should be ok to figure this out down the path.